rclpy/rclpy/context.py

```diff
@@ -123,23 +140,25 @@ def try_shutdown(self):
                 self._call_on_shutdown_callbacks()
                 self._logging_fini()
 
-    def _remove_callback(self, weak_method):
+    def _remove_callback(self, weak_method: WeakMethod[MethodType]) -> None:
         self._callbacks.remove(weak_method)
 
-    def on_shutdown(self, callback: Callable[[], None]):
+    def on_shutdown(self, callback: Callable[[], None]) -> None:
         """Add a callback to be called on shutdown."""
         if not callable(callback):
             raise TypeError('callback should be a callable, got {}', type(callback))
+        if self.__context is None:
+            raise RuntimeError('Context must be initialized before it can be shutdown')
```

Comment:
I think we can skip this check. I think it's fine to add `on_shutdown` callbacks before initializing the context because it would be a way to avoid a race condition where one thread initialized the context, but another thread (or the signal handler) shutdown the context before the `on_shutdown` callbacks could be registered.

Reply:
Were you thinking of something like this? We need to check if `__context` is None otherwise it will error on `self.__context.ok()`

```python
 def on_shutdown(self, callback: Callable[[], None]) -> None:
        """Add a callback to be called on shutdown."""
        if not callable(callback):
            raise TypeError('callback should be a callable, got {}', type(callback))
        if self.__context is None:
            if ismethod(callback):
                self._callbacks.append(WeakMethod(callback, self._remove_callback))
            else:
                self._callbacks.append(callback)
            return

        with self.__context, self._lock:
            if not self.__context.ok():
                callback()
            else:
                if ismethod(callback):
                    self._callbacks.append(WeakMethod(callback, self._remove_callback))
                else:
                    self._callbacks.append(callback)
```

Reply:
Ah yeah I missed that we were holding `with self.__context` when adding the callback.